kv/kvserver: TestSharedLocksBasic failed during spanset assertion

kv/kvserver.TestSharedLocksBasic [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11966302?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11966302?buildTab=artifacts#/) on master @ [2111b61b2d7c789bc03b1e9392062df80c779075](https://github.com/cockroachdb/cockroach/commits/2111b61b2d7c789bc03b1e9392062df80c779075):


```
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch(0xc00b789080, {0xde298b0?, 0xc00caec030}, 0xc00de2c900, {{0xc0073cc6b8, 0x1, 0x8}, {0xc0073cc6c8, 0x1, 0x8}}, ...)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1875 +0x9ac
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges(0xc00b789080, {0xde298b0, 0xc00caec030}, 0xc00de2c900, {{0xc0073cc6b8, 0x1, 0x8}, {0xc0073cc6c8, 0x1, 0x8}}, ...)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1446 +0x573
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send(0xc00b789080, {0xde298b0, 0xc00b08ff80}, 0xc00de2c900)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1066 +0xa25
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnLockGatekeeper).SendLocked(0xc00d66f5b0, {0xde298b0, 0xc00b08ff80}, 0xc00de2c900)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_lock_gatekeeper.go:82 +0x22a
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnMetricRecorder).SendLocked(0xc00d66f578, {0xde298b0, 0xc00b08ff80}, 0x4ace95?)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_metric_recorder.go:47 +0x1b9
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnSpanRefresher).sendLockedWithRefreshAttempts(0xc00d66f498, {0xde298b0, 0xc00b08ff80}, 0xc00de2c900, 0x5)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go:222 +0x2b6
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnSpanRefresher).SendLocked(0xc00d66f498, {0xde298b0, 0xc00b08ff80}, 0xc00de2c900)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go:150 +0x1e7
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnCommitter).SendLocked(0xc00d66f460, {0xde298b0, 0xc00b08ff80}, 0xc00de2c900)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go:146 +0xa06
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnPipeliner).SendLocked(0xc00d66f330, {0xde298b0, 0xc00b08ff80}, 0x1?)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go:295 +0x246
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnSeqNumAllocator).SendLocked(0xc00d66f310, {0xde298b0, 0xc00b08ff80}, 0xc00de2c900)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator.go:114 +0x3a6
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnHeartbeater).SendLocked(0xc00d66f260, {0xde298b0, 0xc00b08ff80}, 0xc00de2c900)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_heartbeater.go:246 +0x7d4
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*TxnCoordSender).Send(0xc00d66f080, {0xde29840, 0xc0001bc008}, 0xc00de2c900)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_coord_sender.go:533 +0xa33
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).sendUsingSender(0xc00cdcb360, {0xde29840, 0xc0001bc008}, 0xc00de2c900, {0x7f1799b530b0, 0xc00d66f080})
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:1090 +0x191
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Send(0xc018ed5600, {0xde29840, 0xc0001bc008}, 0xc00de2c900)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:1218 +0x2e9
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.sendAndFill({0xde29840, 0xc0001bc008}, 0xc00c473b70, 0xc00d66fb80)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:922 +0x253
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Run(0xc018ed5600, {0xde29840, 0xc0001bc008}, 0x8f31940?)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:764 +0xa5
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).scan(0xc018ed5600, {0xde29840, 0xc0001bc008}, {0x8f31940, 0xddc6750}, {0x8f31940, 0xddc6760}, 0x0, 0x9a?, 0x2, ...)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:610 +0x22d
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).ScanForShare(...)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:652
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestSharedLocksBasic.func1(0xc00df54000, 0x1)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_replica_test.go:4991 +0x111
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.RunValues[...].func1()
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:31 +0x56
        	            	  | testing.tRunner(0xc00df54000, 0xc006e641c0)
        	            	  | 	GOROOT/src/testing/testing.go:1576 +0x217
        	            	  | created by testing.(*T).Run
        	            	  | 	GOROOT/src/testing/testing.go:1629 +0x806
        	            	Error types: (1) *errbase.opaqueWrapper (2) *errutil.leafError
        	Test:       	TestSharedLocksBasic/guaranteed-durability=true
=== NAME  TestSharedLocksBasic
    client_replica_test.go:5030: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestSharedLocksBasic2243548617
    --- FAIL: TestSharedLocksBasic/guaranteed-durability=true (0.13s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSharedLocksBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-31888

kv/kvserver.TestSharedLocksBasic [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11982742?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11982742?buildTab=artifacts#/) on master @ [d89e0f438dfd8fdd1e891bd222f88205458966f4](https://github.com/cockroachdb/cockroach/commits/d89e0f438dfd8fdd1e891bd222f88205458966f4):


```
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch(0xc019eab600, {0xde49770?, 0xc00bed1680}, 0xc00db49200, {{0xc0071d9e28, 0x1, 0x8}, {0xc0071d9e38, 0x1, 0x8}}, ...)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1875 +0x9ac
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges(0xc019eab600, {0xde49770, 0xc00bed1680}, 0xc00db49200, {{0xc0071d9e28, 0x1, 0x8}, {0xc0071d9e38, 0x1, 0x8}}, ...)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1446 +0x573
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send(0xc019eab600, {0xde49770, 0xc00bed15f0}, 0xc00db49200)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1066 +0xa25
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnLockGatekeeper).SendLocked(0xc00b4576b0, {0xde49770, 0xc00bed15f0}, 0xc00db49200)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_lock_gatekeeper.go:82 +0x22a
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnMetricRecorder).SendLocked(0xc00b457678, {0xde49770, 0xc00bed15f0}, 0x4ace95?)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_metric_recorder.go:47 +0x1b9
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnSpanRefresher).sendLockedWithRefreshAttempts(0xc00b457598, {0xde49770, 0xc00bed15f0}, 0xc00db49200, 0x5)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go:222 +0x2b6
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnSpanRefresher).SendLocked(0xc00b457598, {0xde49770, 0xc00bed15f0}, 0xc00db49200)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go:150 +0x1e7
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnCommitter).SendLocked(0xc00b457560, {0xde49770, 0xc00bed15f0}, 0xc00db49200)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go:146 +0xa06
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnPipeliner).SendLocked(0xc00b457430, {0xde49770, 0xc00bed15f0}, 0x1?)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go:295 +0x246
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnSeqNumAllocator).SendLocked(0xc00b457410, {0xde49770, 0xc00bed15f0}, 0xc00db49200)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator.go:114 +0x3a6
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnHeartbeater).SendLocked(0xc00b457360, {0xde49770, 0xc00bed15f0}, 0xc00db49200)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_heartbeater.go:246 +0x7d4
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*TxnCoordSender).Send(0xc00b457180, {0xde49700, 0xc0001bc008}, 0xc00db49200)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_coord_sender.go:533 +0xa33
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).sendUsingSender(0xc004df2fa0, {0xde49700, 0xc0001bc008}, 0xc00db49200, {0x7f1c419c01f0, 0xc00b457180})
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:1090 +0x191
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Send(0xc002f02210, {0xde49700, 0xc0001bc008}, 0xc00db49200)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:1218 +0x2e9
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.sendAndFill({0xde49700, 0xc0001bc008}, 0xc009f2fb70, 0xc00db6a000)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:922 +0x253
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Run(0xc002f02210, {0xde49700, 0xc0001bc008}, 0x8f485e0?)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:764 +0xa5
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).scan(0xc002f02210, {0xde49700, 0xc0001bc008}, {0x8f485e0, 0xdde65e0}, {0x8f485e0, 0xdde65f0}, 0x0, 0x9a?, 0x2, ...)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:610 +0x22d
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).ScanForShare(...)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:652
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestSharedLocksBasic.func1(0xc00a9d1ba0, 0x1)
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_replica_test.go:4994 +0x111
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.RunValues[...].func1()
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:31 +0x56
        	            	  | testing.tRunner(0xc00a9d1ba0, 0xc008935480)
        	            	  | 	GOROOT/src/testing/testing.go:1576 +0x217
        	            	  | created by testing.(*T).Run
        	            	  | 	GOROOT/src/testing/testing.go:1629 +0x806
        	            	Error types: (1) *errbase.opaqueWrapper (2) *errutil.leafError
        	Test:       	TestSharedLocksBasic/guaranteed-durability=true
=== NAME  TestSharedLocksBasic
    client_replica_test.go:5033: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestSharedLocksBasic3599415893
    --- FAIL: TestSharedLocksBasic/guaranteed-durability=true (0.15s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSharedLocksBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/spanset/BUILD.bazel
+++ b/pkg/kv/kvserver/spanset/BUILD.bazel
@@ -33,6 +33,7 @@ go_test(
     embed = [":spanset"],
     deps = [
         "//pkg/keys",
+        "//pkg/kv/kvserver/concurrency/lock",
         "//pkg/roachpb",
         "//pkg/storage",
         "//pkg/testutils",

--- a/pkg/kv/kvserver/spanset/batch.go
+++ b/pkg/kv/kvserver/spanset/batch.go
@@ -891,6 +891,19 @@ func addLockTableSpans(spans *SpanSet) *SpanSet {
 		if span.EndKey != nil {
 			ltEndKey, _ = keys.LockTableSingleKey(span.EndKey, nil)
 		}
+		if sa == SpanReadOnly && span.Timestamp == hlc.MaxTimestamp {
+			// Shared lock acquisition uses a read-only latch access with
+			// the maximum timestamp. This gives it sufficient isolation to
+			// write to the lock table without having to declare a write
+			// latch and be serialized with other shared lock acquisitions.
+			// For details, see DefaultDeclareIsolatedKeys.
+			//
+			// For the sake of this function, we consider this to be strong
+			// enough to declare write access to the lock table. This could
+			// be made cleaner if latch spans operated on locking strengths
+			// instead of read/write access.
+			sa = SpanReadWrite
+		}
 		withLocks.AddNonMVCC(sa, roachpb.Span{Key: ltKey, EndKey: ltEndKey})
 	})
 	return withLocks

--- a/pkg/kv/kvserver/spanset/batch_test.go
+++ b/pkg/kv/kvserver/spanset/batch_test.go
@@ -15,6 +15,7 @@ import (
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/storage"
@@ -46,13 +47,31 @@ func TestReadWriterDeclareLockTable(t *testing.T) {
 		},
 	}
 	for fnName, fn := range fns {
-		for _, sa := range []spanset.SpanAccess{spanset.SpanReadOnly, spanset.SpanReadWrite} {
+		for _, str := range []lock.Strength{lock.None, lock.Shared, lock.Exclusive, lock.Intent} {
 			for _, mvcc := range []bool{false, true} {
-				t.Run(fmt.Sprintf("%s,access=%s,mvcc=%t", fnName, sa, mvcc), func(t *testing.T) {
+				if !mvcc && (str == lock.Shared || str == lock.Exclusive) {
+					// Invalid combination.
+					continue
+				}
+				t.Run(fmt.Sprintf("%s,strength=%s,mvcc=%t", fnName, str, mvcc), func(t *testing.T) {
 					span := roachpb.Span{Key: startKey, EndKey: endKey}
+					var sa spanset.SpanAccess
+					var latchTs hlc.Timestamp
+					switch str {
+					case lock.None:
+						sa, latchTs = spanset.SpanReadOnly, ts
+					case lock.Shared:
+						sa, latchTs = spanset.SpanReadOnly, hlc.MaxTimestamp
+					case lock.Exclusive:
+						sa, latchTs = spanset.SpanReadWrite, ts
+					case lock.Intent:
+						sa, latchTs = spanset.SpanReadWrite, ts
+					default:
+						t.Fatal("unexpected")
+					}
 					ss := spanset.New()
 					if mvcc {
-						ss.AddMVCC(sa, span, ts)
+						ss.AddMVCC(sa, span, latchTs)
 					} else {
 						ss.AddNonMVCC(sa, span)
 					}
@@ -64,10 +83,10 @@ func TestReadWriterDeclareLockTable(t *testing.T) {
 					require.Error(t, rw.MVCCIterate(ltEndKey, ltEndKey.Next(), storage.MVCCKeyIterKind, storage.IterKeyTypePointsOnly, nil))
 
 					err := rw.PutUnversioned(ltStartKey, []byte("value"))
-					if sa == spanset.SpanReadWrite {
-						require.NoError(t, err)
-					} else {
+					if str == lock.None {
 						require.Error(t, err)
+					} else {
+						require.NoError(t, err)
 					}
 					require.Error(t, rw.PutUnversioned(ltEndKey, []byte("value")))
 				})

